core/builtin.py

```diff
@@ -483,9 +483,24 @@ def Cd(argv, mem):
   return 0
 
 
-def PrintDirStack(dir_stack):
-  if dir_stack:
-    print(' '.join(dir_stack))
+WITH_PREFIX = 1
+WITHOUT_PREFIX = 2
+SINGLE_LINE = 3
+def PrintDirStack(dir_stack, mode = 3):
+  dirs = [os.getcwd()]
+  dirs.extend(dir_stack)
+  if mode == 1:
+    for i, entry in enumerate(dirs):
+      print('%2d  %s' % (i, entry))
+    sys.stdout.flush()
+  elif mode == 2:
+    for entry in dirs:
+      print(entry)
+    sys.stdout.flush()
+  elif mode == 3:
+    if dirs:
```

Comment:
dirs will always be non-empty, so you can remove this.

Also sys.stdout.flush() should go at the end of the function.